File: NetworkServer/src/protocol/UdpPacketCodec.cpp
#include "protocol/UdpPacketCodec.h"
#include "protocol/NetworkByteOrder.h"

#include <algorithm>
#include <cstring>
#include <stdexcept>

namespace
{
    void WriteNetwork16(
        std::vector<std::uint8_t>& destination,
        const std::size_t offset,
        const std::uint16_t value
    )
    {
        const std::uint16_t networkValue =
            tdr::protocol::HostToNetwork16(value);

        std::memcpy(
            destination.data() + offset,
            &networkValue,
            sizeof(networkValue)
        );
    }

    void WriteNetwork32(
        std::vector<std::uint8_t>& destination,
        const std::size_t offset,
        const std::uint32_t value
    )
    {
        const std::uint32_t networkValue =
            tdr::protocol::HostToNetwork32(value);

        std::memcpy(
            destination.data() + offset,
            &networkValue,
            sizeof(networkValue)
        );
    }

    std::uint16_t ReadNetwork16(
        const std::uint8_t* const source,
        const std::size_t offset
    )
    {
        std::uint16_t networkValue{};

        std::memcpy(
            &networkValue,
            source + offset,
            sizeof(networkValue)
        );

        return tdr::protocol::NetworkToHost16(networkValue);
    }

    std::uint32_t ReadNetwork32(
        const std::uint8_t* const source,
        const std::size_t offset
    )
    {
        std::uint32_t networkValue{};

        std::memcpy(
            &networkValue,
            source + offset,
            sizeof(networkValue)
        );

        return tdr::protocol::NetworkToHost32(networkValue);
    }
}

namespace tdr::protocol
{
    UdpPacketDecodeException::UdpPacketDecodeException(
        const UdpPacketError code,
        const std::string& message
    )
        : std::runtime_error(message),
        code_(code)
    {
    }

    UdpPacketError
        UdpPacketDecodeException::Code() const noexcept
    {
        return code_;
    }

    std::vector<std::uint8_t> UdpPacketCodec::Encode(
        const UdpMessageHeader& header,
        const std::vector<std::uint8_t>& payload
    )
    {
        std::vector<std::uint8_t> packet(
            kUdpMessageHeaderSize + payload.size()
        );

        WriteNetwork32(
            packet,
            kUdpMagicOffset,
            header.magic
        );

        WriteNetwork16(
            packet,
            kUdpVersionOffset,
            header.version
        );

        WriteNetwork16(
            packet,
            kUdpMessageTypeOffset,
            static_cast<std::uint16_t>(header.type)
        );

        std::copy(
            header.sessionToken.begin(),
            header.sessionToken.end(),
            packet.begin() + kUdpSessionTokenOffset
        );

        WriteNetwork32(
            packet,
            kUdpPlayerIdOffset,
            header.playerId
        );

        WriteNetwork32(
            packet,
            kUdpSequenceOffset,
            header.sequence
        );

        std::copy(
            payload.begin(),
            payload.end(),
            packet.begin() + kUdpMessageHeaderSize
        );

        return packet;
    }

    DecodedUdpPacket UdpPacketCodec::Decode(
        const std::uint8_t* const data,
        const std::size_t size
    )
    {
        if (data == nullptr)
        {
            throw std::invalid_argument(
                "UDP datagram data cannot be null."
            );
        }

        if (size < kUdpMessageHeaderSize)
        {
            throw std::invalid_argument(
                "UDP datagram is smaller than its header."
            );
        }

        DecodedUdpPacket packet;

        packet.header.magic =
            ReadNetwork32(data, kUdpMagicOffset);

        if (packet.header.magic != kUdpProtocolMagic)
        {
            throw UdpPacketDecodeException(
                UdpPacketError::InvalidMagic,
                "Invalid UDP protocol magic."
            );
        }

        packet.header.version =
            ReadNetwork16(data, kUdpVersionOffset);

        if (packet.header.version != kUdpProtocolVersion)
        {
            throw UdpPacketDecodeException(
                UdpPacketError::UnsupportedVersion,
                "Unsupported UDP protocol version."
            );
        }

        packet.header.type =
            static_cast<MessageType>(
                ReadNetwork16(data, kUdpMessageTypeOffset)
                );

        if (!IsKnownMessageType(packet.header.type))
        {
            throw UdpPacketDecodeException(
                UdpPacketError::UnknownMessageType,
                "Unknown UDP message type."
            );
        }

        if (!IsUdpMessageType(packet.header.type))
        {
            throw UdpPacketDecodeException(
                UdpPacketError::NonUdpMessageType,
                "Message type is not valid for UDP."
            );
        }

        std::copy(
            data + kUdpSessionTokenOffset,
            data + kUdpSessionTokenOffset
            + kUdpSessionTokenSize,
            packet.header.sessionToken.begin()
        );

        packet.header.playerId =
            ReadNetwork32(data, kUdpPlayerIdOffset);

        packet.header.sequence =
            ReadNetwork32(data, kUdpSequenceOffset);

        packet.payload.assign(
            data + kUdpMessageHeaderSize,
            data + size
        );

        return packet;
    }
}

File: Assets/Scripts/Networking/Protocol/UdpPacketCodec.cs
using System;

namespace TopDownRoguelike.Networking.Protocol
{
    public enum UdpPacketError : byte
    {
        InvalidMagic,
        UnsupportedVersion,
        UnknownMessageType,
        NonUdpMessageType
    }

    public sealed class UdpPacketDecodeException
        : Exception
    {
        public UdpPacketDecodeException(
            UdpPacketError code,
            string message)
            : base(message)
        {
            Code = code;
        }

        public UdpPacketError Code { get; }
    }

    public readonly struct UdpMessageHeader
    {
        public UdpMessageHeader(
            MessageType type,
            byte[] sessionToken,
            uint playerId,
            uint sequence)
        {
            if (sessionToken == null)
            {
                throw new ArgumentNullException(
                    nameof(sessionToken));
            }

            if (sessionToken.Length !=
                UdpPacketCodec.SessionTokenSize)
            {
                throw new ArgumentException(
                    "UDP session token must contain 16 bytes.",
                    nameof(sessionToken));
            }

            Type = type;

            SessionToken =
                (byte[])sessionToken.Clone();

            PlayerId = playerId;
            Sequence = sequence;
        }

        public MessageType Type { get; }

        public byte[] SessionToken { get; }

        public uint PlayerId { get; }

        public uint Sequence { get; }
    }

    public readonly struct DecodedUdpPacket
    {
        public DecodedUdpPacket(
            UdpMessageHeader header,
            byte[] payload)
        {
            Header = header;

            Payload = payload
                ?? throw new ArgumentNullException(
                    nameof(payload));
        }

        public UdpMessageHeader Header { get; }

        public byte[] Payload { get; }
    }

    public static class UdpPacketCodec
    {
        public const uint ProtocolMagic =
            0x54445255u;

        public const ushort ProtocolVersion =
            1;

        public const int MagicOffset =
            0;

        public const int VersionOffset =
            4;

        public const int MessageTypeOffset =
            6;

        public const int SessionTokenOffset =
            8;

        public const int SessionTokenSize =
            16;

        public const int PlayerIdOffset =
            24;

        public const int SequenceOffset =
            28;

        public const int MessageHeaderSize =
            32;

        public static byte[] Encode(
            UdpMessageHeader header,
            byte[] payload)
        {
            if (payload == null)
            {
                throw new ArgumentNullException(
                    nameof(payload));
            }

            ValidateUdpMessageType(
                header.Type,
                nameof(header));

            if (header.SessionToken == null ||
                header.SessionToken.Length !=
                SessionTokenSize)
            {
                throw new ArgumentException(
                    "UDP session token must contain 16 bytes.",
                    nameof(header));
            }

            var packet =
                new byte[
                    MessageHeaderSize +
                    payload.Length];

            PacketCodec.WriteNetworkUInt32(
                packet,
                MagicOffset,
                ProtocolMagic);

            PacketCodec.WriteNetworkUInt16(
                packet,
                VersionOffset,
                ProtocolVersion);

            PacketCodec.WriteNetworkUInt16(
                packet,
                MessageTypeOffset,
                (ushort)header.Type);

            Buffer.BlockCopy(
                header.SessionToken,
                0,
                packet,
                SessionTokenOffset,
                SessionTokenSize);

            PacketCodec.WriteNetworkUInt32(
                packet,
                PlayerIdOffset,
                header.PlayerId);

            PacketCodec.WriteNetworkUInt32(
                packet,
                SequenceOffset,
                header.Sequence);

            if (payload.Length > 0)
            {
                Buffer.BlockCopy(
                    payload,
                    0,
                    packet,
                    MessageHeaderSize,
                    payload.Length);
            }

            return packet;
        }

        public static DecodedUdpPacket Decode(
            byte[] datagram)
        {
            if (datagram == null)
            {
                throw new ArgumentNullException(
                    nameof(datagram));
            }

            if (datagram.Length <
                MessageHeaderSize)
            {
                throw new ArgumentException(
                    "UDP datagram is smaller than its header.",
                    nameof(datagram));
            }

            uint magic =
                PacketCodec.ReadNetworkUInt32(
                    datagram,
                    MagicOffset);

            if (magic != ProtocolMagic)
            {
                throw new UdpPacketDecodeException(
                    UdpPacketError.InvalidMagic,
                    "Invalid UDP protocol magic.");
            }

            ushort version =
                PacketCodec.ReadNetworkUInt16(
                    datagram,
                    VersionOffset);

            if (version != ProtocolVersion)
            {
                throw new UdpPacketDecodeException(
                    UdpPacketError.UnsupportedVersion,
                    "Unsupported UDP protocol version.");
            }

            var messageType =
                (MessageType)
                PacketCodec.ReadNetworkUInt16(
                    datagram,
                    MessageTypeOffset);

            if (!Enum.IsDefined(
                typeof(MessageType),
                messageType))
            {
                throw new UdpPacketDecodeException(
                    UdpPacketError.UnknownMessageType,
                    "Unknown UDP message type.");
            }

            if (!IsUdpMessageType(messageType))
            {
                throw new UdpPacketDecodeException(
                    UdpPacketError.NonUdpMessageType,
                    "Message type is not valid for UDP.");
            }

            var sessionToken =
                new byte[SessionTokenSize];

            Buffer.BlockCopy(
                datagram,
                SessionTokenOffset,
                sessionToken,
                0,
                SessionTokenSize);

            uint playerId =
                PacketCodec.ReadNetworkUInt32(
                    datagram,
                    PlayerIdOffset);

            uint sequence =
                PacketCodec.ReadNetworkUInt32(
                    datagram,
                    SequenceOffset);

            int payloadSize =
                datagram.Length -
                MessageHeaderSize;

            var payload =
                new byte[payloadSize];

            if (payloadSize > 0)
            {
                Buffer.BlockCopy(
                    datagram,
                    MessageHeaderSize,
                    payload,
                    0,
                    payloadSize);
            }

            var header =
                new UdpMessageHeader(
                    messageType,
                    sessionToken,
                    playerId,
                    sequence);

            return new DecodedUdpPacket(
                header,
                payload);
        }

        private static void ValidateUdpMessageType(
            MessageType messageType,
            string parameterName)
        {
            if (!Enum.IsDefined(
                typeof(MessageType),
                messageType))
            {
                throw new ArgumentException(
                    "Unknown UDP message type.",
                    parameterName);
            }

            if (!IsUdpMessageType(messageType))
            {
                throw new ArgumentException(
                    "Message type is not valid for UDP.",
                    parameterName);
            }
        }

        private static bool IsUdpMessageType(
            MessageType messageType)
        {
            switch (messageType)
            {
                case MessageType.UdpBindRequest:
                case MessageType.UdpBindAccepted:
                case MessageType.UdpPing:
                case MessageType.UdpPong:
                case MessageType.PlayerInput:
                case MessageType.PlayerStateSnapshot:
                case MessageType.PlayerShotEvent:
                case MessageType.PlayerShotgunEvent:
                    return true;

                default:
                    return false;
            }
        }
    }
}